tests/ignite/metrics/test_hsic.py

```diff
@@ -0,0 +1,184 @@
+        # non-linearly dependent
+        N = 200
+        b = 20
+        x = torch.randn(N, 5)
+        y = x @ torch.normal(0.0, torch.pi, size=(5, 3))
+        y = (
+            torch.stack([torch.sin(y[:, 0]), torch.cos(y[:, 1]), torch.exp(y[:, 2])], dim=1)
+            + torch.randn_like(y) * 1e-4
+        )
+
+    return x, y, b
+
+
+@pytest.mark.parametrize("n_times", range(3))
+@pytest.mark.parametrize("sigma_x", [-1.0, 1.0])
+@pytest.mark.parametrize("sigma_y", [-1.0, 1.0])
+def test_compute(n_times, sigma_x: float, sigma_y: float, test_case: Tuple[Tensor, Tensor, int]):
+    x, y, batch_size = test_case
+
+    hsic = HSIC(sigma_x=sigma_x, sigma_y=sigma_y)
+
+    hsic.reset()
+
+    np_hsic_sum = 0.0
+    n_iters = y.shape[0] // batch_size
+    for i in range(n_iters):
+        idx = i * batch_size
+        x_batch = x[idx : idx + batch_size]
+        y_batch = y[idx : idx + batch_size]
+
+        hsic.update((x_batch, y_batch))
+        np_hsic_sum += np_hsic(x_batch, y_batch, sigma_x, sigma_y)
+    np_res = np_hsic_sum / n_iters
+
+    assert isinstance(hsic.compute(), float)
+    assert pytest.approx(np_res, abs=2e-5) == hsic.compute()
+
+
+def test_accumulator_detached():
+    hsic = HSIC()
+
+    x = torch.rand(10, 10, dtype=torch.float)
+    y = torch.rand(10, 10, dtype=torch.float)
+    hsic.update((x, y))
+
+    assert not hsic._sum_of_hsic.requires_grad
+
+
+@pytest.mark.usefixtures("distributed")
+class TestDistributed:
+    @pytest.mark.parametrize("sigma_x", [-1.0, 1.0])
+    @pytest.mark.parametrize("sigma_y", [-1.0, 1.0])
+    def test_integration(self, sigma_x: float, sigma_y: float):
+        tol = 2e-5
+        n_iters = 100
+        batch_size = 20
+        n_dims_x = 100
+        n_dims_y = 50
+
+        rank = idist.get_rank()
+        torch.manual_seed(12 + rank)
+
+        device = idist.device()
+        metric_devices = [torch.device("cpu")]
+        if device.type != "xla":
+            metric_devices.append(device)
+
+        for metric_device in metric_devices:
+            x = torch.randn((n_iters * batch_size, n_dims_x)).float().to(device)
+
+            lin = nn.Linear(n_dims_x, n_dims_y).to(device)
+            y = torch.sin(lin(x) * 100) + torch.randn(n_iters * batch_size, n_dims_y) * 1e-4
+
+            def data_loader(i):
+                return x[i * batch_size : (i + 1) * batch_size], y[i * batch_size : (i + 1) * batch_size]
+
+            engine = Engine(lambda e, i: data_loader(i))
+
+            m = HSIC(sigma_x=sigma_x, sigma_y=sigma_y, device=metric_device)
+            m.attach(engine, "hsic")
+
+            data = list(range(n_iters))
+            engine.run(data=data, max_epochs=1)
+
+            assert "hsic" in engine.state.metrics
+            res = engine.state.metrics["hsic"]
+
+            np_res = 0.0
+            for i in range(n_iters):
+                x_batch, y_batch = data_loader(i)
+                np_res += np_hsic(x_batch, y_batch)
+            np_res = np_res / n_iters
```

Comment:
I think this expected value computation is not correct. `test_integration` function in distributed tests is run by multiple processes, so `HSIC` will compute average HSIC over all processes. On the other hand, `np_res` is computed per process and thus wont match.

Reply:
Here is working test code:
```python
    @pytest.mark.parametrize("sigma_x", [-1.0, 1.0])
    @pytest.mark.parametrize("sigma_y", [-1.0, 1.0])
    def test_integration(self, sigma_x: float, sigma_y: float):
        tol = 2e-5
        n_iters = 100
        batch_size = 20
        n_dims_x = 100
        n_dims_y = 50

        rank = idist.get_rank()
        torch.manual_seed(12 + rank)

        device = idist.device()
        metric_devices = [torch.device("cpu")]
        if device.type != "xla":
            metric_devices.append(device)

        for metric_device in metric_devices:
            x = torch.randn((n_iters * batch_size, n_dims_x)).float().to(device)

            lin = nn.Linear(n_dims_x, n_dims_y).to(device)
            y = torch.sin(lin(x) * 100) + torch.randn(n_iters * batch_size, n_dims_y) * 1e-4

            def data_loader(i, input_x, input_y):
                return input_x[i * batch_size : (i + 1) * batch_size], input_y[i * batch_size : (i + 1) * batch_size]

            engine = Engine(lambda e, i: data_loader(i, x, y))

            m = HSIC(sigma_x=sigma_x, sigma_y=sigma_y, device=metric_device)
            m.attach(engine, "hsic")

            data = list(range(n_iters))
            engine.run(data=data, max_epochs=1)

            assert "hsic" in engine.state.metrics
            res = engine.state.metrics["hsic"]

            x = idist.all_gather(x)
            y = idist.all_gather(y)
            total_n_iters = idist.all_reduce(n_iters)

            np_res = 0.0
            for i in range(total_n_iters):
                x_batch, y_batch = data_loader(i, x, y)
                np_res += np_hsic(x_batch, y_batch, sigma_x, sigma_y)

            expected_hsic = np_res / total_n_iters
            assert pytest.approx(expected_hsic, abs=tol) == res
```